internal/net/grpc/stats/stats.go

```diff
@@ -0,0 +1,394 @@
+	CGV1
+	CGV2
+)
+
+const (
+	cgroupBasePath = "/sys/fs/cgroup"
+)
+
+// CgroupMetrics holds raw values directly read from cgroup files
+type CgroupMetrics struct {
+	Mode CgroupMode
+
+	// Raw values from cgroup files
+	MemUsageBytes uint64
+	MemLimitBytes uint64 // 0 means unlimited
+	CPUUsageNano  uint64
+	CPUQuotaUs    uint64 // 0 means unlimited
+	CPUPeriodUs   uint64 // 0 if unknown
+}
+
+// CgroupStats holds calculated resource usage statistics ready for use
+type CgroupStats struct {
+	CpuLimitCores    float64
+	CpuUsageCores    float64
+	MemoryLimitBytes uint64
+	MemoryUsageBytes uint64
+}
+
+func Register(srv *grpc.Server) {
+	ssrv := &server{}
+	statspb.RegisterStatsServer(srv, ssrv)
+}
+
+type server struct {
+	statspb.UnimplementedStatsServer
+}
+
+func (s *server) ResourceStats(
+	ctx context.Context, _ *payload.Empty,
+) (stats *payload.Info_ResourceStats, err error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "unknown"
+	}
+
+	ip := net.LoadLocalIP()
+	if ip == "" {
+		ip = "unknown"
+	}
+
+	cgroupStats, err := measureCgroupStats()
+	if err != nil {
+		log.Debugf("failed to get cgroup stats: %v", err)
+		stats = &payload.Info_ResourceStats{
+			Name: hostname,
+			Ip:   ip,
+			CgroupStats: &payload.Info_CgroupStats{
+				CpuLimitCores:    0,
+				CpuUsageCores:    0,
+				MemoryLimitBytes: 0,
+				MemoryUsageBytes: 0,
+			},
+		}
+		err = nil
+		return stats, err
+	}
+
+	stats = &payload.Info_ResourceStats{
+		Name: hostname,
+		Ip:   ip,
+		CgroupStats: &payload.Info_CgroupStats{
+			CpuLimitCores:    cgroupStats.CpuLimitCores,
+			CpuUsageCores:    cgroupStats.CpuUsageCores,
+			MemoryLimitBytes: cgroupStats.MemoryLimitBytes,
+			MemoryUsageBytes: cgroupStats.MemoryUsageBytes,
+		},
+	}
+
+	return stats, nil
+}
+
+// measureCgroupStats orchestrates the process of sampling and calculating cgroup statistics.
+func measureCgroupStats() (stats *CgroupStats, err error) {
+	// First sample: Read initial metrics from cgroup files (includes cumulative CPU usage)
+	m1, err := readCgroupMetrics()
+	if err != nil {
+		return nil, errors.ErrCgroupFirstSampleFailed(err)
+	}
+	t1 := time.Now()
+
+	// Wait 100ms to allow meaningful CPU usage accumulation
+	time.Sleep(100 * time.Millisecond)
+
+	// Second sample: Read metrics again to calculate CPU usage rate
+	m2, err := readCgroupMetrics()
+	if err != nil {
+		return nil, errors.ErrCgroupSecondSampleFailed(err)
+	}
+	t2 := time.Now()
+
+	// Calculate CPU usage rate from cumulative values: (usage2 - usage1) / time_delta
+	stats = calculateCpuUsageCores(m1, m2, t2.Sub(t1))
+
+	return stats, nil
+}
+
+// readCgroupMetrics reads raw memory & CPU metrics depending on cgroup mode
+func readCgroupMetrics() (metrics *CgroupMetrics, err error) {
+	mode := detectCgroupMode()
+	switch mode {
+	case CGV2:
+		metrics, err = readCgroupV2Metrics()
+	case CGV1:
+		metrics, err = readCgroupV1Metrics()
+	default:
+		err = errors.ErrCgroupModeDetectionFailed()
+	}
+
+	return metrics, err
```

Comment:
using early return would be better
```suggestion
	switch detectCgroupMode() {
	case CGV2:
		return readCgroupV2Metrics()
	case CGV1:
		return readCgroupV1Metrics()
	default:
		return nil, errors.ErrCgroupModeDetectionFailed()
	}
```